zkevm-circuits/src/table.rs

```diff
@@ -0,0 +1,874 @@
+            |mut region| {
+                let mut offset = 0;
+                for column in self.columns() {
+                    region.assign_advice(
+                        || "keccak table all-zero row",
+                        column,
+                        offset,
+                        || Ok(F::zero()),
+                    )?;
+                }
+                offset += 1;
+
+                let keccak_table_columns = self.columns();
+                for input in inputs.clone() {
+                    for row in Self::assignments(input, randomness) {
+                        // let mut column_index = 0;
+                        for (column, value) in keccak_table_columns.iter().zip_eq(row) {
+                            region.assign_advice(
+                                || format!("keccak table row {}", offset),
+                                *column,
+                                offset,
+                                || Ok(value),
+                            )?;
+                        }
+                        offset += 1;
+                    }
+                }
+                Ok(())
+            },
+        )
+    }
+}
+
+impl TableColumns for KeccakTable {
+    fn columns(&self) -> Vec<Column<Advice>> {
+        vec![
+            self.is_enabled,
+            self.input_rlc,
+            self.input_len,
+            self.output_rlc,
+        ]
+    }
+}
+
+/// Copy Table, used to verify copies of byte chunks between Memory, Bytecode,
+/// TxLogs and TxCallData.
+#[derive(Clone, Copy, Debug)]
+pub struct CopyTable {
+    /// Whether the row is the first read-write pair for a copy event.
+    pub is_first: Column<Advice>,
+    /// The relevant ID for the read-write row, represented as a random linear
+    /// combination. The ID may be one of the below:
+    /// 1. Call ID/Caller ID for CopyDataType::Memory
+    /// 2. RLC encoding of bytecode hash for CopyDataType::Bytecode
+    /// 3. Transaction ID for CopyDataType::TxCalldata, CopyDataType::TxLog
+    pub id: Column<Advice>,
+    /// The source/destination address for this copy step.  Can be memory
+    /// address, byte index in the bytecode, tx call data, and tx log data.
+    pub addr: Column<Advice>,
+    /// The end of the source buffer for the copy event.  Any data read from an
+    /// address greater than or equal to this value will be 0.
+    pub src_addr_end: Column<Advice>,
+    /// The number of bytes left to be copied.
+    pub bytes_left: Column<Advice>,
+    /// The associated read-write counter for this row.
+    pub rw_counter: Column<Advice>,
+    /// Decrementing counter denoting reverse read-write counter.
+    pub rwc_inc_left: Column<Advice>,
+    /// Binary chip to constrain the copy table conditionally depending on the
+    /// current row's tag, whether it is Bytecode, Memory, TxCalldata or
+    /// TxLog.
+    pub tag: BinaryNumberConfig<CopyDataType, 3>,
+}
+
+impl CopyTable {
+    /// Construct a new KeccakTable
```

Comment:
```suggestion
    /// Construct a new CopyTable
```

Reply:
Fixed in https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/622/commits/7955480c55c5be39fc7efb1710ca2dbf58e34ec8